### PR TITLE
fix: Fix panic when rendering a `Paragraph` out of bounds

### PR DESCRIPTION
**Pull request by jwodder**
Fixes #1667.

### CODE REVIEW

**Review by orhun — APPROVED** [review]

File: ratatui-widgets/src/paragraph.rs [hunk 1]
```diff
@@ -477,7 +477,11 @@ fn render_line(wrapped: &WrappedLine<'_, '_>, area: Rect, buf: &mut Buffer, y: u
         // Make sure to overwrite any previous character with a space (rather than a zero-width)
         let symbol = if symbol.is_empty() { " " } else { symbol };
         let position = Position::new(area.left() + x, area.top() + y);
-        buf[position].set_symbol(symbol).set_style(*style);
+        if let Some(cell) = buf.cell_mut(position) {
+            cell.set_symbol(symbol).set_style(*style);
+        } else {
+            break;
+        }
```
Comment (joshka) [inline ratatui-widgets/src/paragraph.rs:484]: Another approach that might also be worth considering for this is pre-trimming the area function in the render function. (e.g. `let area = buf.area.intersection(area);` But I think that would change how paragraphs that are partially out of the area are rendered, which may not be desirable. Longer note about this in the test below.

File: ratatui-widgets/src/paragraph.rs [hunk 2]
```diff
@@ -1203,4 +1207,18 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_render_paragraph_out_of_bounds() {
+        let paragraph = Paragraph::new("Beyond the pale");
+        for area in [
+            Rect::new(0, 5, 15, 1),
+            Rect::new(20, 0, 15, 1),
+            Rect::new(20, 5, 15, 1),
+        ] {
```
Comment (joshka) [inline ratatui-widgets/src/paragraph.rs:1218]: These areas are fully outside of the buffer area. So the tests here are perfectly correct. I wonder though if there's a second edge case (as mentioned above) where the render area is partially outside the buffer area. There's a couple of ways that this could be intended to render:
1. crop the text outside of the buffer area
2. restrict the paragraph to the buffer area so that it fills correctly

Both of these seem like they'd be useful in some cases. The first one would be useful where you have moving text which you're fine with rendering partially off screen. The second is where you have text which you want to render fully regardless of the screen bounds. So I'm not 100% sure what the default behavior should be here. Either way, it's worth capturing what our current behavior is as a characterization test and then we can work out what it should be later.

Comment (jwodder) [reply ratatui-widgets/src/paragraph.rs:1218]: @joshka So... are you asking for further changes to this PR, or are these things for other people to do in later PRs?

Comment (j-g00da) [reply ratatui-widgets/src/paragraph.rs]: I think 1 seems like a good default behavior (so current implementation), one can always do an intersect if they want it to be restricted.

> Either way, it's worth capturing what our current behavior is as a characterization test and then we can work out what it should be later.

Agree

**Comment by j-g00da** [comment]
@jwodder would you like to continue working on this PR, or would you prefer that I take it over?

**Comment by jwodder** [comment]
@j-g00da You take it over.

**Comment by joshka** [comment]
whoops - beat you to it :)
